Hello psalm

kesselb opened the pull request:
Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Hello psalm

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Add stubs for phpseclib

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Add template functions

for some reason they are not loaded by default.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Add psalm as composer dependency

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Fix: UndefinedFunction Function x does not exist in templates

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Exclude stubs from code style check

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Suppress warning for $_ and $l

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

MorrisJobke commented:
How nice is this? 😍 👍

kesselb committed:
Remove directory/file element.

It will suppress any error of that kind in the file. That was not intended.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Supress UndefinedGlobalVariable for register_command.php

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb commented:
Yep. PHPStorm is going to add support for Psalm soon: https://blog.jetbrains.com/phpstorm/2020/07/phpstan-and-psalm-support-coming-to-phpstorm/. 

I'm not sure yet how to continue with this PR. I fixed some errors or added rules to suppress them (like the global template variables `$_` and `$l`). But there are still more than 1000 warnings.

There are some false positives left. But others warning are legit. To continue and merge this PR we could set a baseline (https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#using-a-baseline-file) with the current errors. With the risk that no one will fix the existing issues ;)

MorrisJobke commented:
> There are some false positives left. But others warning are legit. To continue and merge this PR we could set a baseline (https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#using-a-baseline-file) with the current errors. With the risk that no one will fix the existing issues ;)

I guess that is the only way to go for now.

MorrisJobke commented on psalm.xml:
```diff
@@ -49,5 +49,13 @@
+				<directory name="apps/**/templates"/>
+				<directory name="core/templates"/>
+			</errorLevel>
+		</UndefinedGlobalVariable>
```
Seems to not work properly:

```
/github/workspace/apps/user_ldap/templates/part.wizardcontrols.php:9:11:error - UndefinedGlobalVariable: Cannot find referenced variable $l in global scope
1769
/github/workspace/apps/user_ldap/templates/part.wizardcontrols.php:11:19:error - UndefinedFunction: Function link_to_docs does not exist
1770
/github/workspace/apps/user_ldap/templates/part.wizardcontrols.php:13:35:error - UndefinedFunction: Function image_path does not exist
```

Plus the functions from https://github.com/nextcloud/server/blob/master/lib/private/legacy/template/functions.php should be added in the `< UndefinedFunction>` element.

kesselb replied:
> Seems to not work properly:

I added the feature to psalm and documented it wrong :rofl:

kesselb replied:
Fixed.

MorrisJobke replied:
😆 https://github.com/vimeo/psalm/pull/3990

kesselb committed:
Fix variable name for referenced variable.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Suppress warning for template functions.

https://github.com/nextcloud/server/blob/6e8e34fef920a073118c22111f0f31eb3b3a91dc/lib/private/legacy/template/functions.php
Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

MorrisJobke commented:
> > There are some false positives left. But others warning are legit. To continue and merge this PR we could set a baseline (https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#using-a-baseline-file) with the current errors. With the risk that no one will fix the existing issues ;)
> 
> I guess that is the only way to go for now.

Mind to do this now and then we could merge this :)

skjnldsv commented:
I thought it was a dashboard widget to display Psalm quotes :grin:

MorrisJobke committed:
Add psalm baseline

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
Add a check for fixes in the psalm baseline

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
Add stubs for gd, intl, IntlChar, ldap, redis_cluster and xsl

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
Using stubs found some inconsistencies between the stub and the actual implementation:

https://github.com/JetBrains/phpstorm-stubs/pull/893

I will open more PRs there.

MorrisJobke committed:
Add patches for stubs

* https://github.com/JetBrains/phpstorm-stubs/pull/893
* https://github.com/JetBrains/phpstorm-stubs/pull/894
* https://github.com/JetBrains/phpstorm-stubs/pull/895
* https://github.com/JetBrains/phpstorm-stubs/pull/896

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
Check only the baseline.xml and exclude the psalm.xml from the file check

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
Here it is - the final rebase that squashed all the fixups into the proper commits. It now also warns about not checked in fixes (so if we fix something it also states that the baseline should be updated).

kesselb commented on build/files-checker.php:
```diff
@@ -70,6 +70,7 @@
 	'ocs',
 	'package-lock.json',
 	'package.json',
+	'psalm.xml',
```
We don't need psalm baseline here?

MorrisJobke replied:
No - because it's in the `build/` folder which is deleted already. Only new files in the root need to be checked if the release script should either delete them or keep them. I went for the "delete the psalm.xml in the release" way. :)
